core/Enclave/key_factory.cpp

```diff
@@ -0,0 +1,301 @@
+#include "openssl/ec.h"
+#include "openssl/pem.h"
+#include "openssl/bio.h"
+
+#include "key_factory.h"
+#include "key_operation.h"
+
+#define SGX_DOMAIN_KEY_SIZE     16
+
+#define RSA_2048_KEY_BITS   2048
+#define RSA_3072_KEY_BITS   3072
+#define RSA_4096_KEY_BITS   4096
+
+#define RSA_2048_PUBLIC_KEY_PEM_SIZE    426
+#define RSA_2048_PRIVATE_KEY_PEM_SIZE    1679
+
+#define RSA_3072_PUBLIC_KEY_PEM_SIZE    625
+#define RSA_3072_PRIVATE_KEY_PEM_SIZE    2484
+
+#define RSA_4096_PUBLIC_KEY_PEM_SIZE    775
+#define RSA_4096_PRIVATE_KEY_PEM_SIZE    3247
+
+#define ECC_PUBLIC_KEY_PEM_SIZE     178
+#define ECC_PRIVATE_KEY_PEM_SIZE    227
+#define ECC_MAX_PLAINTEXT_SIZE      256
+
+sgx_aes_gcm_128bit_key_t g_domain_key = {0};
+
+static uint32_t ehsm_get_key_pem_size(const uint32_t keyspec)
+{
+    switch (keyspec)
+    {
+        case EH_RSA_2048:
+            return RSA_2048_PUBLIC_KEY_PEM_SIZE + RSA_2048_PRIVATE_KEY_PEM_SIZE;
+        case EH_RSA_3072:
+            return RSA_3072_PUBLIC_KEY_PEM_SIZE + RSA_3072_PRIVATE_KEY_PEM_SIZE;
+        case EH_RSA_4096:
+            return RSA_4096_PUBLIC_KEY_PEM_SIZE + RSA_4096_PRIVATE_KEY_PEM_SIZE;
+        case EH_EC_P256:
+        case EH_SM2:
+            return ECC_PUBLIC_KEY_PEM_SIZE + ECC_PRIVATE_KEY_PEM_SIZE;
+        default:
+            return UINT32_MAX;
+    }
+}
+
+static uint32_t ehsm_calc_gcm_data_size(const uint32_t aad_size, const uint32_t plaintext_size)
+{
+    if (aad_size > UINT32_MAX - sizeof(sgx_aes_gcm_data_ex_t))
+        return UINT32_MAX;
+
+    if (plaintext_size > UINT32_MAX - sizeof(sgx_aes_gcm_data_ex_t))
+        return UINT32_MAX;
+
+    if (aad_size > UINT32_MAX - plaintext_size)
+        return UINT32_MAX;
+
+    if (sizeof(sgx_aes_gcm_data_ex_t) > UINT32_MAX - plaintext_size - aad_size)
+        return UINT32_MAX;
+
+    return (aad_size + plaintext_size + sizeof(sgx_aes_gcm_data_ex_t));
+}
+
+static uint32_t ehsm_get_public_key_pem_size(const uint32_t keyspec)
+{
+    switch(keyspec)
+    {
+        case EH_RSA_2048:
+            return RSA_2048_PUBLIC_KEY_PEM_SIZE;
+        case EH_RSA_3072:
+            return RSA_3072_PUBLIC_KEY_PEM_SIZE;
+        case EH_RSA_4096:
+            return RSA_4096_PUBLIC_KEY_PEM_SIZE;
+        case EH_EC_P256:
+        case EH_SM2:
+            return ECC_PUBLIC_KEY_PEM_SIZE;
+        default:
+            return 0;
+    }
+}
+
+sgx_status_t ehsm_create_keyblob(const uint32_t plaintext_size, const uint8_t *plaintext,
+                                        const uint32_t aad_size, const uint8_t *aad,
+                                        const uint32_t gcm_data_size, sgx_aes_gcm_data_ex_t *gcm_data)
+{
+    uint32_t real_aad_size = aad_size;
+    if (NULL == aad)
+        real_aad_size = 0;
+
+    sgx_status_t ret = sgx_read_rand(gcm_data->iv, sizeof(gcm_data->iv));
+    if (ret != SGX_SUCCESS) {
+        printf("error generating iv.\n");
+        return ret;
+    }
+
+    ret = sgx_rijndael128GCM_encrypt(&g_domain_key,
```

Comment:
same here, domankey will upgrade to 256bits

Reply:
copy